server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -1140,25 +1248,26 @@ public <T> List<T> readOptionalList(final Writeable.Reader<T> reader) throws IOE
      * @return the list of strings
      * @throws IOException if an I/O exception occurs reading the list
      */
-    public List<String> readOptionalStringList() throws IOException {
-        return readOptionalList(StreamInput::readString);
+    public List<String> readOptionalStringCollectionAsList() throws IOException {
+        return readOptionalCollectionAsList(StreamInput::readString);
     }
 
     /**
-     * Reads a set of objects. If the returned set contains any entries it will be mutable. If it is empty it might be immutable.
+     * Reads a set of objects. The set is expected to have been written using {@link StreamOutput#writeCollection(Collection)}}. If the
```

Comment:
Is it worth giving some info on the behaviour if you write a List with duplicates but read a Set, now the coupling between read & write is looser? Will it dedup, throw exceptions? Or is that undefined behaviour?

Reply:
I think I'd rather not pass judgement on that here. It's almost certainly a mistake to do that, but I don't think we need to define it.

Reply:
In the spirit of progress over perfection I'd prefer to defer any further Javadoc improvements until later PRs (which you're welcome to open). We can definitely say a lot more here, although I'm not sure more detail will really solve any significant problems in practice.